spdx/parsers/jsonyamlxml.py

```diff
@@ -1486,6 +1486,39 @@ def parse_pkg_chksum(self, pkg_chksum):
             self.value_error("PKG_CHECKSUM", pkg_chksum)
 
 
+def unflatten_document(document):
+    """
+    Inverse of spdx.writers.jsonyamlxml.flatten_document
+    """
+    files_by_id = {}
+    if "files" in document:
+        for f in document.pop("files"):
+            f["name"] = f.pop("fileName")
+            # XXX must downstream rely on "sha1" property?
+            for checksum in f["checksums"]:
+                if checksum["algorithm"] == "SHA1":
+                    f["sha1"] = checksum["checksumValue"]
+                    break
+            if "licenseInfoInFiles" in f:
+                f["licenseInfoFromFiles"] = f.pop("licenseInfoInFiles")
+            files_by_id[f["SPDXID"]] = f
+
+    packages = document.pop("packages")
+    for package in packages:
+        if "hasFiles" in package:
+            package["files"] = [{
+                "File": files_by_id[spdxid]} for spdxid in package["hasFiles"]
+            ]
+        # XXX must downstream rely on "sha1" property?
+        for checksum in package.get("checksums", []):
+            if checksum["algorithm"] == "SHA1":
+                package["sha1"] = checksum["checksumValue"]
+                break
+
+    document["documentDescribes"] = [{ "Package": package} for package in packages ]
```

Comment:
This does not comply with the json schema, which just has an array of spdx ids here: https://github.com/spdx/spdx-spec/blob/development/v2.3.1/schemas/spdx-schema.json#L219-L226

Even if it's just a temporary internal data structure, I find this highly confusing

Reply:
I don't see the point of flattening the structure in this case though. The only instance where the value is accessed is [this part where the packages are parsed](https://github.com/spdx/tools-python/blob/76272c47035475126bf3252a2d72d100eff372d7/spdx/parsers/jsonyamlxml.py#L1688-L1689), and this can be done just as well without modifying the data structure first. In fact, it [is already done this way](https://github.com/spdx/tools-python/blob/76272c47035475126bf3252a2d72d100eff372d7/spdx/parsers/jsonyamlxml.py#L1704-L1705) (although the fact that this logic is technically duplicated should be fixed).

If resolving links becomes annoying, I would find it more natural to add a utility method that handles it instead of introducing an additional intermediate data structure. This will come with a performance impact if called many times, but I would be surprised if this was an issue.

Additional remark: The current structure only considers files that are included in packages. To my understanding, files should also be allowed to be independent of any package (this might be related to the python tools still being out of date though, it seems like snippets are not considered at all, for example).

Reply:
One additional comment: My understanding so far is that in a json document, any `SPDX-Document DESCRIBES SPDX-Package` relationship would be omitted and instead the package id would appear in `documentDescribes` (similar for files and snippets). This would imply that during parsing, we have to create new relationships for elements of `documentDesccribes`. I don't see that happening anywhere.

Relevant discussion: https://github.com/spdx/tools-python/issues/242. Let's see if I'm wrong here :sweat_smile: